feat: Adding more natural sounding voices to phone calls

AlaricWhitney opened the pull request:
- [x] Identified the issue which this PR solves.
- [x] Read the [**CONTRIBUTING**](https://github.com/target/goalert/blob/master/CONTRIBUTING.md) document.
- [x] Code builds clean without any errors or warnings.
- [x] Added appropriate tests for any new functionality.
- [x] All new and existing tests passed.
- [x] Added comments in the code, where necessary.
- [x] Ran `make check` to catch common errors. Fixed any that came up.

**Description:**
This PR updates the application to have be able to set the Twilio Phone call language and Text To Speech engine. 
- Currently the default language is "" (which Twilio interprets as `us-EN`).  This PR allows you to adjust it to whatever you need.
- Currently the default TTS engine is Twilio Basic Man, which sounds like a robot.  This PR allows you to adjust it to whatever you need.
- The greeting is also adjusted to make it more natural sounding by adjusting the verbiage, and adding a pause between the greeting and the alert message.

**Which issue(s) this PR fixes:**
Fixes #2574 
Fixes #2576

**Out of Scope:**
NA

**Screenshots:**
NA

**Describe any introduced user-facing changes:**
By default, nothing changes for the user, however the user does have the new ability to change the TTS voice engine and the language for phone calls.

**Describe any introduced API changes:**
NA

**Additional Info:**
NA

mastercactapus commented on config/config.go:
```diff
@@ -360,6 +363,22 @@ func (cfg Config) ApplicationName() string {
 }
 
+// VoiceName will return the Twilio.VoiceName value
+func (cfg Config) VoiceName() string {
+	if cfg.Twilio.VoiceName == "" {
+		return "Polly.Joanna-Neural"
+	}
+	return cfg.Twilio.VoiceName
+}
+
+// VoiceLanguage will return the Twilio.VoiceLanguage value
+func (cfg Config) VoiceLanguage() string {
+	if cfg.Twilio.VoiceLanguage == "" {
+		return "en-US"
+	}
+	return cfg.Twilio.VoiceLanguage
+}
```
The default should be empty so that the voice config continues to be whatever is configured in the Twilio account unless a value is set

AlaricWhitney replied:
adjusted as per your request

mastercactapus commented on config/config.go:
```diff
@@ -508,6 +527,14 @@ func (cfg Config) Validate() error {
 		)
 	}
 
+	if cfg.Twilio.VoiceName != "" {
+		err = validate.Many(
+			err,
+			validate.ASCII("Twilio.VoiceName", cfg.Twilio.VoiceName, 1, 50),
+			validate.ASCII("Twilio.VoiceLanguage", cfg.Twilio.VoiceLanguage, 0, 10),
+		)
+	}
```
We can simplify this since neither is required

```suggestion
	err = validate.Many(
		err,
		validate.ASCII("Twilio.VoiceName", cfg.Twilio.VoiceName, 0, 50),
		validate.ASCII("Twilio.VoiceLanguage", cfg.Twilio.VoiceLanguage, 0, 10),
	)
```

AlaricWhitney replied:
adjusted as per your request.  Technically you require a language if you specify a voice, which I adjusted the check to validate.

mastercactapus commented on notification/twilio/twiml.go:
```diff
@@ -134,7 +142,7 @@ func (t *twiMLResponse) sendResponse() {
 	}
 	for _, s := range t.say {
-		io.WriteString(t.w, `<Say><prosody rate="slow">`)
+		io.WriteString(t.w, fmt.Sprintf(`<Say voice="%s" language="%s"><prosody rate="slow">`, t.voiceName, t.voiceLanguage))
```
Only set the voice and/or language if the admin has explicitly set the new config option. An admin should expect no changes to the outgoing configuration from applying an update; so we need to introduce the feature in a way that is completely opt-in.

AlaricWhitney replied:
adjusted as per your request.  Now it will only be set if the config option is set, otherwise nothing changes.

mastercactapus commented on notification/twilio/voice.go:
```diff
@@ -187,7 +187,7 @@ func (v *Voice) Send(ctx context.Context, msg notification.Message) (*notificati
 	}
 
-	prefix := fmt.Sprintf("This is %s", cfg.ApplicationName())
+	prefix := fmt.Sprintf("Hello! This is %s. <break strength=\"x-strong\" time=\"700ms\"/>", cfg.ApplicationName())
```
We will need to update all the messages, too, with this change.

Currently, the prefix is spoken as part of the introduction, "This is {application} with alert notifications {details}" we'll want to address the text in the switch statement below.

Maybe drop a message in the chatroom about the proposal to ensure buy-in. My understanding is you're proposing we restructure voice messages from "{Appname & message type} {details} {options}" to "{appname} {details} {options}" or maybe the intent is "{appname} {type} {details} {options}"?

In any case, it's worth a few minutes to chat about the how and why so we can review and ensure all calls are consistent and all reviewers know what to look for.

AlaricWhitney replied:
Actually, this is just a minor wording an vocal pacing update.  The structure for phone calls will still remain `{Appname} {details} {options}`, and that is untouched.  The current Greeting is `This is GoAlert {details} {options}` which is all run together and sounds like it's all one long sentence.  This change simple adjusts the greeting to `Hello! This is GoAlert. <700ms pause> {details} {options}`, which adds a natural vocal separation between the greeting, and the message details.

mastercactapus replied:
If you look at line 196, for example, we are appending the reason for the call to the prefix, which places it after the pause as a broken sentence.

The current greeting is:
`This is GoAlert with an alert notification. {details} {options}`

With this change it is:
`Hello! This is GoAlert. <700ms pause> with an alert notification. {details} {options}`

I think it should go before the pause, or be rephrased to be a complete sentence.

For example:
`Hello! This is GoAlert with an alert notification. <700ms pause> {details} {options}`

AlaricWhitney replied:
Updated all of the pauses to do `Hello! This is GoAlert with an alert notification. <700ms pause> {details} {options}`

mastercactapus commented on notification/twilio/voice.go:
```diff
@@ -572,10 +581,12 @@ func (v *Voice) ServeAlert(w http.ResponseWriter, req *http.Request) {
 
 	// See Twilio Request Parameter documentation at
 	// https://www.twilio.com/docs/api/twiml/twilio_request#synchronous
 	resp := newTwiMLResponse(w)
```
Let's add `ctx` as an arg to `newTwiMLResponse` and have it read in the config internally so we don't need to track calling `AddVoiceOptions` at every call site

AlaricWhitney replied:
adjusted as per your request

mastercactapus commented on notification/twilio/twiml.go:
```diff
@@ -133,8 +142,18 @@ func (t *twiMLResponse) sendResponse() {
 	}
+
+	// add optional voice and/or language options
+	voiceAndLanuage := ""
+	if t.voiceName != "" && t.voiceLanguage != "" {
+		// language is required with a voice
+		voiceAndLanuage = fmt.Sprintf(` voice="%s" language="%s"`, t.voiceName, t.voiceLanguage)
+	} else if t.voiceLanguage != "" {
+		// language can be set without a voice
+		voiceAndLanuage = fmt.Sprintf(` language="%s"`, t.voiceLanguage)
+	}
```
nit: we can simplify this and avoid the `else` by just doing an `if` for each, the name-then-lang requirement is already enforced by config validation.

```go
var opts string
if voice {
  opts += ` voice="%s"`
}
if lang {
  opts += ` lang="%s"`
}
```

**However,** the bigger issue here is that we need to escape the values otherwise this is vulnerable to TwiML injection through the UI.

We could lift and use the escaping code from the XML lib [here](https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/encoding/xml/xml.go;drc=8d57f4dcef5d69a0a3f807afaa9625018569010b;l=1961), but I'm thinking it might be a good sign that it's time to use the proper XML library to build this.

AlaricWhitney replied:
Updated to use the xml parser.

mastercactapus commented on notification/twilio/twiml_test.go:
```diff
@@ -1,19 +1,23 @@
 
 func TestTwiMLResponse(t *testing.T) {
 	t.Run("hangup", func(t *testing.T) {
+		mockConfig := config.Config{}
```
Use `var` for empty declarations, also since it's just a regular config object (not a fake/mock), I think we can call these `cfg`
```suggestion
		var cfg config.Config
```

mastercactapus commented on config/config.go:
```diff
@@ -97,6 +97,9 @@ type Config struct {
 	Twilio struct {
 		Enable bool `public:"true" info:"Enables sending and processing of Voice and SMS messages through the Twilio notification provider."`
 
+		VoiceName     string `public:"true" info:"The Twilio voice to use for Text To Speech for phone calls."`
+		VoiceLanguage string `public:"true" info:"The Twilio voice language to use for Text To Speech for phone calls."`
```
Are we using these in the UI outside of the admin config panel? If not, we can leave the `public` tag off. That is only intended to indicate fields that should be readable by non-admin users.

AlaricWhitney replied:
not really.  removed the public tags.

mastercactapus commented on config/config.go:
```diff
@@ -508,6 +511,14 @@ func (cfg Config) Validate() error {
 		)
 	}
 
+	if cfg.Twilio.VoiceName != "" {
+		err = validate.Many(
+			err,
+			validate.ASCII("Twilio.VoiceName", cfg.Twilio.VoiceName, 1, 50),
+			validate.ASCII("Twilio.VoiceLanguage", cfg.Twilio.VoiceLanguage, 1, 10), // language is required when a voice name is set
+		)
+	}
```
Don't let language escape validation if name is unset

```suggestion
	if cfg.Twilio.VoiceName != "" {
		err = validate.Many(
			err,
			validate.ASCII("Twilio.VoiceName", cfg.Twilio.VoiceName, 1, 50),

			// language is required when a voice name is set
			validate.ASCII("Twilio.VoiceLanguage", cfg.Twilio.VoiceLanguage, 1, 10),
		)
	} else {
		err = validate.Many(
			err,
			validate.ASCII("Twilio.VoiceLanguage", cfg.Twilio.VoiceLanguage, 0, 10),
		)
	}
```

AlaricWhitney replied:
applied.

mastercactapus reviewed (changes requested):
I'm going to look into using the XML encoder to deal with the escaping issue (as a separate PR), I already started some of that with the mocktwilio refactor work, so I think it will be straightforward.

AlaricWhitney commented:
@mastercactapus Cool.  I'll wait until your XML PR #2582 goes in before making more updates to this PR.

mastercactapus commented:
@AlaricWhitney the XML PR has been merged

mastercactapus commented:
@AlaricWhitney, just a note on this, I haven't had a chance to dig in, but it doesn't seem to work with the Polly voices. The `<break>` blocks get read out literally. I think it all needs to be wrapped in `<speak>` or something like that to use the "advanced" syntax (I forget the acronym)

AlaricWhitney commented:
@mastercactapus This is ready for review again.  I've implemented the message to include an extra optional query parameter that indicates where in the string slice to insert pauses.  I've also added unit tests and benchmark tests to show how it works.  I also moved some of the message functionality into their own separate functions to make unit testing easier.

mastercactapus commented on notification/twilio/voice.go:
```diff
@@ -401,6 +357,17 @@ func (v *Voice) getCall(w http.ResponseWriter, req *http.Request) (context.Conte
+
+	// add pause index values
+	var pauseIndexData []int
+	if q.Has(msgParamPause) {
+		for _, val := range q[msgParamPause] {
```
Range works over `nil`, so we can skip the `Has` check

mastercactapus commented on notification/twilio/voice.go:
```diff
@@ -401,6 +357,17 @@ func (v *Voice) getCall(w http.ResponseWriter, req *http.Request) (context.Conte
+	var pauseIndexData []int
+	if q.Has(msgParamPause) {
+		for _, val := range q[msgParamPause] {
+			if i, err := strconv.Atoi(val); err != nil {
```
An error here should reject with a 400 bad request

mastercactapus commented on notification/twilio/voice.go:
```diff
@@ -638,3 +606,100 @@ func (v *Voice) FriendlyValue(ctx context.Context, value string) (string, error)
+	subID := -1
+	switch t := msg.(type) {
+	case notification.AlertBundle:
+		message = fmt.Sprintf("%s with alert notifications. Service '%s' has %d unacknowledged alerts.", prefix, t.ServiceName, t.Count)
+		pauseIndex = append(pauseIndex, len(fmt.Sprintf("%s with alert notifications.", prefix)))
```
Instead of passing these indexes around everywhere, can we handle encoding/decoding at the request and use a string slice everywhere else?

AlaricWhitney replied:
I think it makes more sense to have it directly by the message for readability and comprehension.  If we move the index injection elsewhere, it won't really be obvious what's going on (ie, ensuring that the message text, and the injection are by each other so you can easily keep them in sync).

mastercactapus replied:
Keeping it separate is basically a pseudo-type, like this:
```go
type MultiString struct {
	MsgBody string
	Indexes []int
}
```

While I agree to keep things surface-level and not add unnecessary abstraction, embedding it as two separate fields in an existing struct doesn't help with readability or comprehension over `[]string`, which will track the whole dataset, but without special knowledge or the necessity to keep indexes in sync.

Tracking indexes manually for the entire lifecycle also adds noise at every layer, like tests, where the reader needs to know why it's there, what it's doing, and how to keep it correct.

The separate-indexes-encoding stuff is ONLY relevant when building or parsing the query parameters, so we can keep that complexity deliberate and localized without maintaining it everywhere else.

https://google.github.io/styleguide/go/guide#simplicity
> - Does not assume that you already know what it is doing
> - Does not assume that you can memorize all of the preceding code

https://google.github.io/styleguide/go/guide#least-mechanism
> Where there are several ways to express the same idea, prefer the one that uses the most standard tools.

AlaricWhitney replied:
@mastercactapus Updated to pass around a type rather than multiple inputs as suggested.

mastercactapus commented on notification/twilio/voice.go:
```diff
@@ -471,13 +438,13 @@ func (v *Voice) ServeTest(w http.ResponseWriter, req *http.Request) {
 		fallthrough
 	case "", digitRepeat:
-		resp.Say(call.msgBody)
+		processSayBody(resp, call.msgBody, call.msgPauseIndex)
```
This feels like it should just be a method

AlaricWhitney replied:
I can do it if you want, although I do think since it's voice specific, so I will make it part of the Voice object.

mastercactapus commented on notification/twilio/voice_test.go:
```diff
@@ -1,12 +1,393 @@
+}
+
+func TestProcessSayBody(t *testing.T) {
+	type mockInput struct {
+		resp          *twiMLResponse
+		msgBody       string
+		msgPauseIndex []int
+	}
+
+	testCases := map[string]struct {
+		input    mockInput
+		expected *twiMLResponse
+	}{
```
resp is empty for all these, which adds noise, and specifying the indexes manually seems fragile.

mastercactapus reviewed (changes requested):
Having to manage/track indexes all over the place is going to be error-prone and noisy; we should be using a slice of strings throughout our application code to manage chunks of text.

The request itself can have the slice encoded in a `string` + `[]int` format, but we should parse that as soon as we get the request into a more usable format of `[]string`, and only switch to indexes at the last moment when we build the URL.

Doing so will let us simplify a lot of this.

AlaricWhitney commented:
> Having to manage/track indexes all over the place is going to be error-prone and noisy; we should be using a slice of strings throughout our application code to manage chunks of text.
> 
> The request itself can have the slice encoded in a `string` + `[]int` format, but we should parse that as soon as we get the request into a more usable format of `[]string`, and only switch to indexes at the last moment when we build the URL.
> 
> Doing so will let us simplify a lot of this.

I chose to do it this way to ensure separation of functionality between what each object is supposed to do, and ensure backwards compatibility.  More specifically, I really think that the message injection processing should occur immediately next to where the message is defined.  Indexes are ordered, so it shouldn't be an issue to pass it around a little bit.

mastercactapus commented:
To help move this along, should we break #2574 into a PR on its own while the `<break>` stuff cycles? I think the voice/language stuff has been good for a while, doesn't require any pause/break changes, and is ready to be validated.

It would also help to limit the scope for review.

mastercactapus commented on notification/twilio/voice.go:
```diff
@@ -35,6 +35,11 @@ type Voice struct {
 	r notification.Receiver
 }
 
+type VoiceMultiString struct {
+	Body         string
+	PauseIndexes []int
+}
```
multi-string already exists as `[]string` in Go

mastercactapus commented:
@AlaricWhitney Separating the two bodies of work should help reduce noise and get traction to move things along. Let's pull the config update changes into a separate PR specifically for #2574 and have this one be the greeting/API changes for #2576. That will also help with our release notes in the future.

Since the voice API callback parameters and rendering code need careful review, and the Twilio voice code is already a bit of a mess in master, it is more expensive to validate. Since we didn't have granular validation for this code, reviewing this PR is more complex. I like that your changes add some unit testing that has been missing, and I think we can break it up a little differently to make it more straightforward.

Let's chat/pair on this to reduce spin, at a high level, I'm thinking:
- To remove a bunch of nested conditionals and simplify, switch to using `[]string`
- To help the readability & reviewability of the tests, find a way to simplify the inputs/outputs for each test case (e.g., maybe changing `buildMessage` slightly)
- See if we can do something about `getCall` to make it testable, or at least a helper func that _is_

AlaricWhitney commented:
@mastercactapus Made the recommended updates to remove the pauses, and only include the voice settings.  ready for review again.

mastercactapus commented:
It looks like the `buildMessage`/refactor work was still left in. It might be cleaner/easier to open a PR from commit 612c1e2f3cf4c8fd0eb113c9ed4d162a4127946d, as that looks like the last one before the pause/break/refactor work started happening.

IIRC the only changes needed from that commit are to remove the XML break since we'll be handling that in the next PR(s), and it looked like some of the tests still had empty struct declarations instead of var.

It will be much easier to validate the functional voice changes against code that already has the rework & testing in place, so for that and to keep the PRs small, let's merge the voice refactor work as one PR on its own and then lastly the API changes for the `<break>` pause at the end.

AlaricWhitney commented:
@mastercactapus I decided to leave that element in based on our offline discussion on breaking up this large function into smaller functions to make unit testing easier.  I can leave it out of this PR if you want, but I do feel that it's an overall benefit to this application if left in, as you can now easily unit test the functional element to ensure that nothing is broken by my changes.

AlaricWhitney commented:
For the sake of getting this moving forward, I've removed the buildMessage enhancement for better testability from this.  I'll add it into the another PR.

tony-tvu reviewed (commented):
Everything is working fine but I'm seeing these logs `fatal: No names found, cannot describe anything.` which I'm not seeing when running on master branch. Is anyone else seeing this on their on when running on this branch?

<img width="868" alt="Screenshot 2023-01-25 at 10 45 15 AM" src="https://user-images.githubusercontent.com/52386267/214625066-6d79bfd4-35c3-462c-9b5c-288263736dac.png">

Forfold commented:
Testing this now and not running into that @tony-tvu 🤔

Forfold commented on config/config.go:
```diff
@@ -432,6 +435,8 @@ func (cfg Config) Validate() error {
 		validateKey("Twilio.AuthToken", cfg.Twilio.AuthToken),
 		validateKey("Twilio.AlternateAuthToken", cfg.Twilio.AlternateAuthToken),
+		validate.ASCII("Twilio.VoiceName", cfg.Twilio.VoiceName, 0, 50),
+		validate.ASCII("Twilio.VoiceLanguage", cfg.Twilio.VoiceLanguage, 0, 10),
```
placeholder text on these could be useful. I wasn't totally sure if `english` would work, then I saw a test later on during the review using `en-US`

AlaricWhitney replied:
@Forfold I didn't see any other examples of default values being used, so by doing so would deviate from the current config patterns.  If these values aren't set (like how it is currently today), it'll use the default setting within Twilio itself, so these don't need to be set at all.

Forfold commented on config/config.go:
```diff
@@ -432,6 +435,8 @@ func (cfg Config) Validate() error {
 		validateKey("Twilio.AccountSID", cfg.Twilio.AccountSID),
 		validateKey("Twilio.AuthToken", cfg.Twilio.AuthToken),
 		validateKey("Twilio.AlternateAuthToken", cfg.Twilio.AlternateAuthToken),
+		validate.ASCII("Twilio.VoiceName", cfg.Twilio.VoiceName, 0, 50),
```
Could we link to a list of names in the description? Or have the default name used filled in

AlaricWhitney replied:
Having the default voice filled in would deviate from the config pattern of this app.  If it's not set (like how it is today) it will default to whatever the Twilio account has set to default.

I'll add the link to the voices in a comment here.

tony-tvu reviewed (approved):
I was able to functionally test this with a real phone call and change the voices. Hopefully the issue with `build | fatal: No names found, cannot describe anything` logs resolves itself when running on GoAlert's repo.
